java/vector/src/main/java/org/apache/arrow/vector/stream/MessageSerializer.java

```diff
@@ -106,37 +97,22 @@ public static ArrowBlock serialize(WriteChannel out, ArrowRecordBatch batch)
       throws IOException {
     long start = out.getCurrentPosition();
     int bodyLength = batch.computeBodyLength();
-    ByteBuffer metadata = WriteChannel.serialize(batch);
-
-    int messageLength = 4 + metadata.remaining() + bodyLength;
-    ByteBuffer serializedHeader =
-        serializeHeader(MessageHeader.RecordBatch, messageLength);
-
-    // Compute the required alignment. This is not a great way to do it. The issue is
-    // that we need to know the message size to serialize the message header but the
-    // size depends on the alignment, which depends on the message header.
-    // This will serialize the header again with the updated size alignment adjusted.
-    // TODO: We really just want sizeof(MessageHeader) from the serializeHeader() above.
-    // Is there a way to do this?
-    long bufferOffset = start + 4 + serializedHeader.remaining() + 4 + metadata.remaining();
-    if (bufferOffset % 8 != 0) {
-      messageLength += 8 - bufferOffset % 8;
-      serializedHeader = serializeHeader(MessageHeader.RecordBatch, messageLength);
-    }
 
-    // Write message header.
-    out.writeIntLittleEndian(serializedHeader.remaining());
-    out.write(serializedHeader);
+    FlatBufferBuilder builder = new FlatBufferBuilder();
+    int batchOffset = batch.writeTo(builder);
+
+    ByteBuffer serializedMessage = serializeMessage(builder, MessageHeader.RecordBatch,
+        batchOffset, bodyLength);
 
-    // Write batch header. with the 4 byte little endian prefix
-    out.writeIntLittleEndian(metadata.remaining());
-    int metadataSize = metadata.remaining();
-    long batchStart = out.getCurrentPosition();
-    out.write(metadata);
+    long metadataStart = out.getCurrentPosition();
+    out.writeIntLittleEndian(serializedMessage.remaining());
+    out.write(serializedMessage);
 
     // Align the output to 8 byte boundary.
     out.align();
 
+    long metadataSize = out.getCurrentPosition() - metadataStart;
```

Comment:
It's significantly simpler if you include the padding in the metadata size. The Flatbuffers library shouldn't have a problem with extra bytes at the end of the buffer (it doesn't in C++)